docs/design/2020-06-24-placement-rules-in-sql.md

```diff
@@ -20,196 +45,193 @@ The scenarios of defining placement rules in SQL include:
 - Increase the replica count of more important data
 - Separate irrelevant data into different stores to improve availability
 
-## Define placement rules
+These scenarios usually fit into one of the following categories:
 
-There are 3 kinds of operations on the placement:
+1. An optimization or availability use case (such as improving data access locality)
+2. A compliance use case (such as ensuring data resides within a geographic region)
 
-* ADD: Add more replicas for one role.
-* ALTER: Override the replica configuration for one role.
-* DROP: Remove the replica configuration for one role.
+This proposal makes some intentional decisions so that both categories of use-cases can be handled correctly. It improves upon the earlier proposal by reducing the risk of misconfiguration which can affect compliance use cases.
 
-They’re all achieved by executing `ALTER TABLE` statements.
+## Detailed Design
 
-### Add placement rules
+### New Syntax Overview
 
-Adding new replicas can be done by one or more `ADD PLACEMENT POLICY` clauses:
+There are two ways to specify placement rules:
 
-```sql
-ALTER TABLE table_name
-	ADD PLACEMENT POLICY CONSTRAINTS=constraints ROLE=role REPLICAS=replicas,
-	...
-```
+1. By assigning `PLACEMENT` directly on a database, table or partition (direct assignment)
+2. By creating a new `PLACEMENT POLICY` and then applying the placement policy to a database, table or partition (placement policy)
 
-This statement indicates TiDB to add replicas for all data of table `table_name`, including indexes.
+Using a `PLACEMENT POLICY` will be recommended for compliance requirements, since it can allow administrators to better keep track of usage. This can be seen as similar to how complex environments will use `ROLES` for management instead of directly assigning privileges to users.
 
-`ADD PLACEMENT POLICY` is just a part of alter options, just like `ADD COLUMN` or `ADD CONSTRAINT`.
+Both syntaxes are considered a [`table_option`](https://dev.mysql.com/doc/refman/8.0/en/alter-table.html), and available in both `CREATE TABLE` and `ALTER TABLE` contexts.
 
-To define multiple roles at once, multiple `ADD PLACEMENT POLICY` clauses can appear in a single `ALTER TABLE` statement, even for the same Raft role. For example:
+#### Direct assignment
+
+Creating a new table with a directly assigned `PLACEMENT` constraints:
 
 ```sql
-ALTER TABLE table_name
-	ADD PLACEMENT POLICY CONSTRAINTS="[+zone=sh]" ROLE=leader REPLICAS=1,
-	ADD PLACEMENT POLICY CONSTRAINTS="[+zone=sh]" ROLE=follower REPLICAS=1
-	ADD PLACEMENT POLICY CONSTRAINTS="[+zone=gz]" ROLE=follower REPLICAS=1;
+CREATE TABLE t1 (
+	id INT NOT NULL PRIMARY KEY,
+	b VARCHAR(100)
+) PLACEMENT CONSTRAINTS="[+zone=sh]" ROLE=leader REPLICAS=1, CONSTRAINTS="[+zone=sh]" ROLE=follower REPLICAS=1, CONSTRAINTS="[+zone=gz]" ROLE=follower REPLICAS=1;
```

Comment:
If it's in the `create table` statement, there are 2 problems:
1. When exporting the table to MySQL by Dumpling, MySQL can't recognize it.
2. When exporting the table to TiDB by Dumpling, the downstream TiDB may do not have the same zones.

Will these be trubles?

Reply:
Yes, they might be. I have this in `Unresolved Questions > Behaviors`. We will have to discuss which is the correct behavior (warn or error). I'm not sure actually, I can see advantages to both.

Reply:
When dumping `CREATE TABLE` info I would expect TiDB specific things to be in a comment, just like the `CLUSTERED` keyword for the primary key:

<pre><span style="background-color:#E4E4E4"><font color="#005F5F">5.7.25-TiDB-v5.2.0-alpha-274-g9abf9d0bb</font></span><span style="background-color:#3A3A3A"><font color="#E4E4E4"></font></span><span style="background-color:#3A3A3A"><font color="#FDF6E3"> 127.0.0.1:4000 </font></span><span style="background-color:#6C6C6C"><font color="#3A3A3A"></font></span><span style="background-color:#6C6C6C"><font color="#FDF6E3"> test </font></span><span style="background-color:#D75F00"><font color="#6C6C6C"></font></span><span style="background-color:#D75F00"><font color="#FDF6E3"> SQL </font></span><span style="background-color:#073642"><font color="#D75F00"> </font></span>create table t1 (id serial primary key);
Query OK, 0 rows affected (0.1621 sec)
<span style="background-color:#E4E4E4"><font color="#005F5F">5.7.25-TiDB-v5.2.0-alpha-274-g9abf9d0bb</font></span><span style="background-color:#3A3A3A"><font color="#E4E4E4"></font></span><span style="background-color:#3A3A3A"><font color="#FDF6E3"> 127.0.0.1:4000 </font></span><span style="background-color:#6C6C6C"><font color="#3A3A3A"></font></span><span style="background-color:#6C6C6C"><font color="#FDF6E3"> test </font></span><span style="background-color:#D75F00"><font color="#6C6C6C"></font></span><span style="background-color:#D75F00"><font color="#FDF6E3"> SQL </font></span><span style="background-color:#073642"><font color="#D75F00"> </font></span>show create table t1\G
*************************** 1. row ***************************
       Table: t1
Create Table: CREATE TABLE `t1` (
  `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */,
  UNIQUE KEY `id` (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
1 row in set (0.0008 sec)
</pre>

Reply:
Yes indeed, this is clarified under "show create table".

The problem is the behavior when it knows how to parse the syntax, but the named `PLACEMENT POLICY` has not been defined. This behavior (warning or error) is what is in question.

Reply:
I think that should be an error.

Reply:
Thanks, that is currently how it is defined. The next question is how dumpers should work around restore if they need to. It is a similar problem to how `SET FOREIGN_KEY_CHECKS=0` is pre-pended by mysqldump etc.

(There is another conversation below about this. It can be an option like this, or a way to change the `SHOW CREATE TABLE` output.)

Reply:
We can make it a special type of comments, just like 

```
 /*T![clustered_index] CLUSTERED */
```

in

```
CREATE TABLE `tbl_0` (
  `col_0` tinytext CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
  `col_1` char(238) COLLATE utf8mb4_unicode_ci NOT NULL,
  `col_2` char(222) CHARACTER SET utf8 COLLATE utf8_unicode_ci NOT NULL,
  `col_3` tinytext COLLATE utf8mb4_general_ci DEFAULT NULL,
  `col_4` text CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
  PRIMARY KEY (`col_2`(3),`col_0`(1)) /*T![clustered_index] CLUSTERED */,
  UNIQUE KEY `idx_1` (`col_1`(1))
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin

```

Reply:
e.g.

```
CREATE TABLE t1 (
	id INT NOT NULL PRIMARY KEY,
	b VARCHAR(100)
) /*T![placement] CONSTRAINTS="[+zone=sh]" ROLE=leader REPLICAS=1, CONSTRAINTS="[+zone=sh]" ROLE=follower REPLICAS=1, CONSTRAINTS="[+zone=gz]" ROLE=follower REPLICAS=1 */
```

Reply:
I have added `PLACEMENT_CHECKS=0` based on `FOREIGN_KEY_CHECKS=0` to the proposal. The default is to check.